airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py

```diff
@@ -163,15 +168,39 @@ def fetch_records(self, job: AsyncJob) -> Iterable[Mapping[str, Any]]:
 
         for url in self.urls_extractor.extract_records(self._polling_job_response_by_id[job.api_job_id()]):
             stream_slice: StreamSlice = StreamSlice(partition={"url": url}, cursor_slice={})
-            # FIXME salesforce will require pagination here
-            response = self.download_requester.send_request(stream_slice=stream_slice)
-            if response:
-                yield from self.record_extractor.extract_records(response)
+            for message in self.download_retriever.read_records({}, stream_slice):
+                if isinstance(message, AirbyteMessage):
+                    if message.type == Type.RECORD:
+                        yield message.record.data  # type: ignore  # message.record won't be None here as the message is a record
+                elif isinstance(message, Record):
+                    yield message.data
+                elif isinstance(message, (dict, Mapping)):
+                    yield message
+                else:
+                    raise TypeError(f"Unknown type `{type(message)}` for message")
```

Comment:
Is there any case when `download_rertiever.read_records` would return `AirbyteMessage` or `Record` instead of `dict`?

My concern is about **performance** because we usually deal with millions of records for async jobs.

Reply:
I think today we only have the SimpleRetrieve which always return an iterable of records. However, the interface allows for anything so it feels like we should play it safe here. Do you have a proposal on how to improve this?

Reply:
> Do you have a proposal on how to improve this?

Yep, we need to get rid of some of the abstractions (like `streamData`) and apply stricter rules for method type outputs. I've done some investigation, but this will require a lot of time to refactor: https://github.com/airbytehq/airbyte-internal-issues/issues/9563

Since we use `SimpleRetriever` only, i think it's safe to bypass record as is, wdyt?

Reply:
> Since we use SimpleRetriever only, i think it's safe to bypass record as is, wdyt?

I think this is true today but I'm afraid it might not always be true. Custom components could act differently. I think the real action item would be to fix our interface if we don't want it to return anything other than records. That being said, this is a bigger change.

Also, I haven't seen a drop in performance for the syncs with the new version so I'm not too worried about that for now. I haven't had very big syncs though. I tried to find one but since streams are incremental, it would require to change the sync mode so I haven't done that.


Would that help if we were to change the order of the `if` for now so that the case for SimpleRetriever be the first?

Reply:
> Would that help if we were to change the order of the if for now so that the case for SimpleRetriever be the first?

yes, sounds good!

Reply:
Updated!